src/main/java/io/cdap/http/NettyHttpService.java

```diff
@@ -301,7 +306,63 @@ public Thread newThread(Runnable r) {
       executor.setKeepAliveTime(execThreadKeepAliveSecs, TimeUnit.SECONDS);
       executor.allowCoreThreadTimeOut(true);
     }
-    return new NonStickyEventExecutorGroup(executor);
+
+    // Returns a EventExecutorGroup that overrides the inEventLoop behavior of the EventExecutor returned by the group.
+    // The inEventLoop will return true if it is calling from a thread in the same thread group, hence was created
+    // from the same executor.
+    // This is to make sure channel events coming from a thread owned by the EventExecutor
+    // will be executed in the same thread instead of submitting a new task.
+    return new ForwardingEventExecutorGroup(new NonStickyEventExecutorGroup(executor)) {
+
+      private final EventExecutorGroup group = this;
+
+      @Override
+      public EventExecutor next() {
+        return wrapEventExecutor(super.next());
+      }
+
+      @Override
+      public Iterator<EventExecutor> iterator() {
+        Iterator<EventExecutor> iterator = super.iterator();
```

Comment:
nit: how about something like `StreamSupport.stream(Spliterators.spliteratorUnknownSize(super.iterator, Spliterator.ORDERED), false).map(this::wrapEventExecutor).iterator()`

Reply:
The one created through StreamSupport doesn't support `remove`, meaning it doesn't preserve all behaviors of the original iterator.